src/Cimpress.Extensions.Http.Caching.InMemory/InMemoryCacheFallbackHandler.cs

```diff
@@ -0,0 +1,136 @@
+        /// <param name="cache">The cache to be used.</param>
+        internal InMemoryCacheFallbackHandler(HttpMessageHandler innerHandler, TimeSpan maxTimeout, TimeSpan cacheDuration, IStatsProvider statsProvider, IMemoryCache cache) : base(innerHandler ?? new HttpClientHandler())
+        {
+            this.StatsProvider = statsProvider ?? new StatsProvider(nameof(InMemoryCacheHandler));
+            this.maxTimeout = maxTimeout;
+            this.cacheDuration = cacheDuration;
+            responseCache = cache ?? new MemoryCache(new MemoryCacheOptions());
+        }
+
+        /// <summary>
+        /// Tries to get the value from the cache, and only calls the delegating handler on cache misses.
+        /// </summary>
+        /// <returns>The HttpResponseMessage from cache, or a newly invoked one.</returns>
+        protected override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+        {
+            // only handle GET methods
+            if (request.Method != HttpMethod.Get)
+            {
+                return await base.SendAsync(request, cancellationToken);
+            }
+
+            var key = request.RequestUri.ToString();
+
+            // start 3 tasks
+            var httpSendTask = base.SendAsync(request, cancellationToken);
+            var timeoutTask = Task.Delay(maxTimeout, cancellationToken);
+            var cacheTask = responseCache.TryGetAsync(key);
+            
+            // ensure the send task completes (or the timeout task, whatever comes first).
+            var firstCompletedTask = await Task.WhenAny(httpSendTask, timeoutTask);
+
+            // timeout occurred
+            if (firstCompletedTask == timeoutTask)
+            {
+                var data = await ExtractCachedResponse(request, cacheTask);
+                if (data != null)
+                {
+                    // update the cache after the http task eventually completes, without awaiting it
+                    var cacheResultTask = httpSendTask.ContinueWith(t => SaveToCache(t.Result, key), TaskContinuationOptions.OnlyOnRanToCompletion);
+
+                    return data;
+                }
+            }
+
+            // we got it from the HTTP data directly, or it wasn't in the cache and got it after the max timeout value; save that result to the cache and return it
+            var response = await httpSendTask;
+            
+            // try to save it to the cache
+            var entry = await SaveToCache(response, key);
+
+            // when successful, return that value
+            if (entry != null)
+            {
+                StatsProvider.ReportCacheMiss(response.StatusCode);
+                return request.PrepareCachedEntry(entry);
+            }
+
+            // when unsuccessful, try to get it from the cache, which was the last successful invocation
```

Comment:
We aren't catching exceptions in this method, will we get here, or will an exception be thrown?